### ISSUE TITLE
feat: add a label field to signers

### ISSUE THREAD

**Issue by varunsrin** [farcasterxyz/hubble#620]
**What is the feature you would like to implement?**

Add an optional 32-byte string label field to SignerAdd messages that allows between 3 and 32 lowercase alphanumeric characters, hyphens and periods. `/[a-z0-9.-]{3,32}/g`

**Why is this feature important?**

Adding labels allows app developers to specify a name for their Signers achieves three outcomes: 

1. Users can easily identify Signers they have created in the past with any wallet
2. Usage of different applications on the network can be approximated
3. Users can choose to see more or less of content from certain applications 

These features are particularly useful if we see different types of clients that are targeting different social use cases. Users may not want to see all the posts from a very noisy "post pictures every day" client in their "serious news" client. 

These identifiers are application provided and user approved and are not entirely trustworthy. While it is possible to implement a more trustworthy signed approach, this will add significant developer and ux complexity and will take time to get right. Instead, we should start with simple labels that unblock problems on the network today and extend the feature later as needed.  

**Will the protocol spec need to be updated??**
Yes, the Signers section needs to add this as an optional field

**Comment by payton** [comment]
Is this a duplicate of #118?

**Comment by varunsrin** [comment]
i like that approach, and lets use #118 for this discussion and leave this issue for signer labels (which we need for reason 1 above anyway)

**Comment by payton** [comment]
Will do. We can discuss getting on the `.well-known` URI registry long-term, too. There were some recent proposals from the identity foundation in regard to DIDs. If there aren't existing standards we can piggyback off of, it could be a fun discussion.

I'm still ramping up on the state of hubs, so perhaps I'm not fully understanding the dire need for a signer label. This feels like a situation where, if we can avoid a protocol change, we should. Especially if this is a temporary measure that will change down the road and protocol changes will only get harder. `Users can easily identify Signers they have created in the past with any wallet` can still be done by defining a well-known endpoint on clients. The caveats are

1. There's no central registry of clients
2. Some signers may not be associated with a public endpoint (this may be a reasonable requirement, though...)

1 can be solved with some quick coordination among clients and easy self-registration.

I don't have strong opinions on this, so I don't expect a response. Just throwing a few last thoughts out there.

**Comment by varunsrin** [comment]
The main reason is that the out of band approach relies on clients to be active and well behaved, which may not always be true. So in such cases where a client shuts down or stops serving data: 

1. Users may not be able to easily tell what signer was for
2. There's no good way to get a feel for which apps are generating messages on behalf of users

On deprecation, since it is an optional property it shouldnt be too hard to remove in the future, should we decide to go down that path.

**Comment by zachterrell57** [comment]
@payton Are you picking up this ticket? If not I was going to

**Comment by payton** [comment]
> @payton Are you picking up this ticket? If not I was going to

@zachterrell57 nope, you can take it!

**Comment by varunsrin** [comment]
@zachterrell57 id wait for @pfletcherhill to weigh in on this before moving forward. anything without a "help wanted" flag is still being debated and may or may not be shipped.

**Comment by pfletcherhill** [comment]
Some comments:
* Note that signer labels and app attribution have decent overlap but not all signers represent apps. I may personally create multiple signers that I use myself and want to be able to differentiate between them (i.e. "PFH signer 1" and "PFH signer 2").
* Given those two examples, I'm not sure we should limit the character set for these. In my mind, they're nicknames rather than usernames. So they're more similar to the name you'd give a bluetooth device, which accept multiple character cases and spaces. My recommendation would be to limit the string to 32 bytes but allow any characters, which is the same validation we have for profile display names.
* Users should be able to update the nickname for a particular signer by sending a more recent SignerAdd message with the same signer pub key and a new label. The signer store shouldn't require any changes to fulfill this requirement, but just stating it to be comprehensive.

**Comment by varunsrin** [comment]
Supportive of these changes, especially since we may want to support non-english unicode labels as well. Let's relax all the requirements and simply make it an optional field of upto 32 bytes.

cc @zachterrell57 i think we're ready to go on this feature if you want to pick it up!

**Comment by varunsrin** [comment]
@sds we should also update our client to add the warpcast label on all new signers generated

**Comment by pfletcherhill** [comment]
@zachterrell57 here are a few more implementation details if it's helpful:
* Update `message.proto` protobuf schema to change `SignerBody` to `SignerAddBody` and add `string name` field to it (I personally like `name` better than `label`)
* Add a new message body called `SignerRemoveBody` that is the same as the current `SignerBody` (i.e. without name) and use it for SignerRemove messages
* Update the validations file in `@farcaster/utils` to validate that the `name` field in SignerAddBody is less than 32 bytes if present. Use `(new TextEncoder).encode("...").length` rather than just `"...".length` to get the length in UTF-8 (we'll be updating other validations to use that method shortly too).

**Comment by pfletcherhill** [comment]
Completed by https://github.com/farcasterxyz/hubble/pull/627
